### PR TITLE
Fixed centering issue with make_cicle_pattern and make_acircle_pattern()

### LINKED ISSUE #18081: gen_pattern.py Pattern Maker has a border issues
##### System information (version)
- OpenCV => 4.2
- Operating System / Platform => Windows 64 Bit
- Compiler => Python3

##### Detailed description

Borders for make_circles_patter() and and make_acircles_pattern() are inconsistent, resulting in an un-centered pattern. 

What a 4x3 asymmetric circle gird with 40mm spacing looks like:
acircleGrid_4x3@40mm

What it should be:
fixed_acircleGrid_4x3@40mm

##### Steps to reproduce
Run gen_pattern.py and observe the result in Gimp (or another svg file viewer)

gen_pattern.py -o circleGrid_13x9@20mm.svg --rows 4 --columns 3 --type circles --square_size 40 -u mm --page_size A4
 --radius_rate 5

## Notes
Have resolved the issue, just need to merge

### PR DESCRIPTION
**Pull request by IanMaquignaz**
Resolves #18081

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under OpenCV (BSD) License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by alalek** [comment]
Thank you for contribution!

As a bugfix this patch should go into [3.4 branch](https://github.com/opencv/opencv/wiki/Branches) first.
We will merge changes from 3.4 into master regularly (weekly/bi-weekly).

So, please:
- change "base" branch of this PR: master => 3.4 (use "Edit" button near PR title)
- rebase your commits from master onto 3.4 branch. For example:
    `git rebase -i --onto upstream/3.4 upstream/master`
    (check list of your commits, save and quit (Esc + "wq" + Enter)
    where `upstream` is configured by following [this GitHub guide](https://help.github.com/articles/configuring-a-remote-for-a-fork/) and fetched (`git fetch upstream`).
- push rebased commits into source branch of your fork (with `--force` option)

Note: no needs to re-open PR, apply changes "inplace".

**Comment by IanMaquignaz** [comment]
I think I re-based correctly? 

As someone pointed out on issue #18081 the asymmetric circles were not offset at 45 degree angles. This has been corrected, though it's my opinion that the description _"CALIB_CB_ASYMMETRIC_GRID uses asymmetric pattern of circles"_ could be expanded to "_CALIB_CB_ASYMMETRIC_GRID uses an asymmetric pattern of circles with intermittent 45 degree offset"_. Unsure where such a documentation change should be applied.  

Also corrected the issue where inches were unsupported due to integer page sizes. Page width and height are now floats. 

Code can be validated using the python snippet below. All measurements should checkout out in Gimp, though note that exporting the SVG file to JPG for loading with cv2.imread(...) skews scaling. This makes comparison of extracted center-points impossible. 

```
#!/usr/bin/env python

# Imports
import cv2


class PatternMaker:
    def __init__(self, cols, rows, square_size, radius_rate, page_width, page_height):
        self.cols = cols
        self.rows = rows
        self.square_size = square_size
        self.radius_rate = radius_rate
        self.width = page_width
        self.height = page_height

    # Generate Object Points
    def make_circles_pattern(self):
        spacing = self.square_size
        r = spacing / self.radius_rate
        pattern_width = ((self.cols - 1.0) * spacing) + (2.0 * r)
        pattern_height = ((self.rows - 1.0) * spacing) + (2.0 * r)
        x_spacing = (self.width - pattern_width) / 2.0
        y_spacing = (self.height - pattern_height) / 2.0
        for x in range(0, self.cols):
            for y in range(0, self.rows):
                print('[', (x * spacing) + x_spacing + r, ',', (y * spacing) + y_spacing + r, ']',)
                # dot = SVG("circle", cx=(x * spacing) + x_spacing + r,
                #           cy=(y * spacing) + y_spacing + r, r=r, fill="black", stroke="none")
                #self.g.append(dot)

    def make_acircles_pattern(self):
        spacing = self.square_size
        r = spacing / self.radius_rate
        pattern_width = ((self.cols - 1.0) * spacing) + (spacing / 2.0) + (2.0 * r)
        pattern_height = ((self.rows - 1.0) * spacing) + (2.0 * r)
        x_spacing = (self.width - pattern_width) / 2.0
        y_spacing = (self.height - pattern_height) / 2.0
        for x in range(0, self.cols):
            for y in range(0, self.rows):
                print('[', (x * spacing) + (0.5 * (y % 2) * spacing) + x_spacing + r, ',', (y * spacing) + y_spacing + r, ']',)
                # dot = SVG("circle", cx=(x * spacing) + (0.5 * (y % 2) * spacing) + x_spacing + r,
                #           cy=(y * spacing) + y_spacing + r, r=r, fill="black", stroke="none")
                #self.g.append(dot)


# Symmetric Ground Truth
columns = 9
rows = 13
square_size = 20
radius_rate = 5
page_width = 210
page_height = 297
pm = PatternMaker(cols=columns, rows=rows, square_size=square_size, radius_rate=radius_rate, page_width=page_width, page_height=page_height)

# Note! Cannot load svg images with opencv. Export with Gimp to JPG or another supported file type
plane = cv2.imread("circleGrid_A4_13x9@20mm.jpg")
cv2.imshow("plane", plane)

# SYMMETRIC TEST

# Test Symmetric
[isFound, centers] = cv2.findCirclesGrid(image=plane, patternSize=(columns, rows), flags=cv2.CALIB_CB_SYMMETRIC_GRID)
markedPlane = cv2.drawChessboardCorners(image=plane, patternSize=(columns, rows), corners=centers, patternWasFound=isFound)

if isFound:
    # Show
    cv2.imshow("plane", plane)

    # Print ground truth centers
    pm.make_circles_pattern()

    # Print Found Centers
    print(centers)
else:
    print('Failed to find centers')

# ASYMMETRIC TEST

# Asymmetric Ground Truth
columns = 4
rows = 11
square_size = 20
radius_rate = 5
page_width = 210
page_height = 297
pm = PatternMaker(cols=columns, rows=rows, square_size=square_size, radius_rate=radius_rate, page_width=page_width, page_height=page_height)

# Note! Cannot load svg images with opencv. Export with Gimp to JPG or another supported file type
aplane = cv2.imread("acircleGrid_A4_11x4@20mm.jpg")
cv2.imshow("aplane", aplane)

[isFound, centers] = cv2.findCirclesGrid(image=aplane, patternSize=(columns, rows), flags=cv2.CALIB_CB_ASYMMETRIC_GRID)
markedPlane = cv2.drawChessboardCorners(image=aplane, patternSize=(columns, rows), corners=centers, patternWasFound=isFound)

if isFound:
    # Show
    cv2.imshow("aplane", aplane)

    # Print ground truth centers
    pm.make_circles_pattern()

    # Print Found Centers
    print(centers)
else:
    print('Failed to find centers')

# End of program
cv2.waitKey()
print('End of Program')
```

**Comment by alalek** [comment]
`findCirclesGrid()` itself may process non-45 degree CALIB_CB_ASYMMETRIC_GRID boards (affine transformation over 45 degree pattern can simulate that case).
Main problem is how it is used after that. For example, in calibration or even solvePnP() calls.

If we change pattern layout, then we need to update "reference" points too.
For example, 45 degree pattern is expected [in this tool](https://github.com/opencv/opencv/blame/4.4.0/apps/interactive-calibration/frameProcessor.cpp#L163-L170).
But it is impossible to fix all User applications (external/private) - we should not break them.

---

So, no need to change documentation.

File: doc/pattern_tools/gen_pattern.py [hunk 1]
```diff
@@ -33,22 +33,29 @@ def __init__(self, cols, rows, output, units, square_size, radius_rate, page_wid
         self.height = page_height
         self.g = SVG("g")  # the svg group container
 
-    def make_circles_pattern(self):
+    def make_acircles_pattern(self):
```
Comment (alalek) [inline doc/pattern_tools/gen_pattern.py:36]: Repair names.

Comment (IanMaquignaz) [reply doc/pattern_tools/gen_pattern.py]: Fixed copy-paste error from experimental. Function names and function order should match the original file.

**Comment by IanMaquignaz** [comment]
> `findCirclesGrid()` itself may process non-45 degree CALIB_CB_ASYMMETRIC_GRID boards (affine transformation over 45 degree pattern can simulate that case).
> If we change pattern layout, then we need to update "reference" points too.

Agreed, pattern and pattern tools should not be changed, but I think the documentation should reflect the currently implicit 45 degree asymmetry. As result of not knowing, I created a denser pattern with a 22.5 degree asymmetry which, as I later discovered in testing, cannot be recovered by `findCirclesGrid()`. 

In my opinion, there is currently a weird dissociation in OpenCV between pattern generation ('make_acircles_pattern()'), pattern recovery (`findCirclesGrid()`) and calibration (calibrateCamera()).  Pattern generation only creates the visual patterns, pattern recovery doesn't accurately describe what it's searching for, and calibration requires source points which are generated by none of the above. This observation entirely describes my current predicament, where I am rewriting metric procam calibration code because whoever wrote it calculated source points as: 
`vector<Point3f> obj;`
`for (int j = 0; j < BOARD_SIZE.height; j++){`
      `for (int k = 0; k < BOARD_SIZE.width; k++){`
              `obj.push_back(Point3f((float)j, (float)k, 0));`
      `}`
`}`

**Comment by IanMaquignaz** [comment]
As a result, I spent much of last week creating a polymorphic c++ class which provides pattern generation, source points and more for chessboards and circle grids.

**Review by alalek — APPROVED** [review]
Looks good to me! Thank you :+1:

**Comment by IanMaquignaz** [comment]
Closes issue #18081

**Comment by IanMaquignaz** [comment]
closes #18081

**Pull request merged**